8305670: Performance regression in LockSupport.unpark with lots of idle threads

dcubed-ojdk opened the pull request:
Address the performance regression in `LockSupport.unpark()` with the following changes:
- Add `FastThreadsListHandle` helper class to threadSMR.[ch]pp for quickly determining if a JavaThread* is protected by the ThreadsListHandle embedded in the new helper object.
- Update `Unsafe_Unpark()` to pass `java_lang_Thread::thread_acquire(thread_oop)` to a new instance of the `FastThreadsListHandle` object and only do the unpark work if the target JavaThread* is protected.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305670](https://bugs.openjdk.org/browse/JDK-8305670): Performance regression in LockSupport.unpark with lots of idle threads


### Reviewers
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [d65ab8e4](https://git.openjdk.org/jdk/pull/13519/files/d65ab8e484db54facc234deaccce3a99473164be)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [86d1b820](https://git.openjdk.org/jdk/pull/13519/files/86d1b820d8869eb319526ac2c9f863c028128c64)


### Contributors
 * Robbin Ehn `<rehn@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13519/head:pull/13519` \
`$ git checkout pull/13519`

Update a local copy of the PR: \
`$ git checkout pull/13519` \
`$ git pull https://git.openjdk.org/jdk.git pull/13519/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13519`

View PR using the GUI difftool: \
`$ git pr show -t 13519`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13519.diff">https://git.openjdk.org/jdk/pull/13519.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13519#issuecomment-1513813954)

dcubed-ojdk commented:
/label add hotspot-runtime
/label remove serviceability
/label remove hotspot

These changes were tested with a new test that's in the bug report and
with Mach5 Tier[1-8] testing.

apangin commented:
@dcubed-ojdk Thank you for working on this. The changes make total sense to me (I'm not a reviewer though).
I tested the patch with the `UnparkRegression` benchmark from the bug report and it now produces 151K calls per second comparing to 70K without the patch. That is a quite good improvement.

robehn reviewed (approved):
Hi Dan, thanks!

Looks good to me.

dholmes-ora commented on src/hotspot/share/prims/unsafe.cpp:
```diff
@@ -783,21 +783,23 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
-        p->unpark();
+    oop thread_oop = JNIHandles::resolve_non_null(jthread);
+    if (java_lang_Thread::thread(thread_oop) != nullptr) {
+      ThreadsListHandle tlh; // Provides memory barrier
```
I don't understand what the comment means - what kind of memory barrier?

robehn replied:
Store side do:
Store to threadslist
Store to eetop

Read side do:
Read eetop
Read threadslist
Read eetop again

We want them to stay in this order.

dcubed-ojdk replied:
Perhaps the comment should change to:
```
      ThreadsListHandle tlh; // Also is a barrier between the two JavaThread* queries.
```

dholmes-ora replied:
I agree we need these ordering constraints, but I'm not clear what is actually guaranteeing them on either the reader or writer side. I imagine there are a number of "barriers" in relation to TLH construction, but as adding to the threadsList is done under a lock (as is setting eetop) then I'm not sure what explicit barriers exist there. Maybe `java_lang_Thread::set_thread` needs to use `release_address_field_put`?

dcubed-ojdk replied:
There are ~76 ThreadsListHandle variables in the code base. These are the only ones with comments:
```
src/hotspot/share/runtime/threadSMR.cpp:    ThreadsListHandle tlh;  // make the current ThreadsList safe for reporting
src/hotspot/share/prims/unsafe.cpp:      ThreadsListHandle tlh; // Provides memory barrier
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
src/hotspot/share/compiler/compileBroker.cpp:        ThreadsListHandle tlh;  // name() depends on the TLH.
```

The new "// Provides memory barrier" is the only one that even talks about barriers.
I'm inclined to delete the comment given that these current lines of code:
```
    if (java_lang_Thread::thread(thread_oop) != nullptr) {
      // Try to capture the live JavaThread in a ThreadsListHandle:
      ThreadsListHandle tlh;
      if (java_lang_Thread::thread(thread_oop) != nullptr) {
        // The still live JavaThread is protected by the ThreadsListHandle.
        JavaThread* thr = nullptr;
```
I think the comments added in the previous comment are enough to make this code clear.
@robehn and @dholmes-ora - are you both good with this?

dholmes-ora replied:
I don't mind if the "memory barrier" comment is there or not. I'm more concerned about whether the actual required ordering constraints are in fact enforced. I think `java_lang_Thread::set_thread` may need to have release semantics (and the matching `acquire` is (implicitly) within the TLH construction).

dcubed-ojdk replied:
@dholmes-ora - I'm sorry that I forgot to reply to this earlier comment in this thread:
https://github.com/openjdk/jdk/pull/13519#discussion_r1171987653

> but as adding to the threadsList is done under a lock (as is setting eetop) 

Adding a new JavaThread* to the main ThreadsList is done under a lock.
The `java_lang_Thread::set_thread()` calls are less clean:
- there's one call from`create_initial_thread()` that doesn't have the Threads_lock, but I also don't think it needs it.
- there's one call from `JavaThread::allocate_threadObj()` which happens during JNI attach that doesn't have the Threads_lock, but I don't think it needs it

- the call from `JavaThread::prepare()` has the Threads_lock
- the call from `JavaThread::start_internal_daemon()` has the Threads_lock

Of course, here's our favorite call from `ensure_join()`:
```
static void ensure_join(JavaThread* thread) {
  // We do not need to grab the Threads_lock, since we are operating on ourself.
<snip>
  OrderAccess::release();
  java_lang_Thread::set_thread(threadObj(), nullptr);
```

Obviously someone thought that call to `java_lang_Thread::set_thread()` needed a "release".

dcubed-ojdk replied:
And I need to address part of this comment:
https://github.com/openjdk/jdk/pull/13519#discussion_r1173249806

> I think java_lang_Thread::set_thread may need to have release semantics
> (and the matching acquire is (implicitly) within the TLH construction).

I agree we should put release semantics in `java_lang_Thread::set_thread()` and
I'll work on that next.

This part is puzzling me:
> the matching acquire is (implicitly) within the TLH construction

TLH construction does nothing with the `JavaThread*` stored in the `java.lang.Thread` object
unless I'm really missing something here. I would think the matching acquire would be in
any call to `java_lang_Thread::thread()`. 

So I also think that `java_lang_Thread::thread()` needs to use `address_field_acquire()`
unless I'm missing something.

dholmes-ora replied:
Yes you are right that makes things simpler and much neater. We end up with two acquires in the path:
```
Read eetop
Read threadslist
Read eetop again
```
but that is okay.

dholmes-ora commented on src/hotspot/share/prims/unsafe.cpp:
```diff
@@ -783,21 +783,23 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
-        HOTSPOT_THREAD_UNPARK((uintptr_t) p);
-        p->unpark();
+    oop thread_oop = JNIHandles::resolve_non_null(jthread);
+    if (java_lang_Thread::thread(thread_oop) != nullptr) {
```
Suggestion for a comment after this line:
```
// We initially have a live JavaThread, so try to capture it in a ThreadsListHandle
```

dcubed-ojdk replied:
I've added:
```
    if (java_lang_Thread::thread(thread_oop) != nullptr) {
      // Try to capture the live JavaThread in a ThreadsListHandle:
      ThreadsListHandle tlh; // Provides memory barrier
```

dholmes-ora commented on src/hotspot/share/prims/unsafe.cpp:
```diff
@@ -783,21 +783,23 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
+    oop thread_oop = JNIHandles::resolve_non_null(jthread);
+    if (java_lang_Thread::thread(thread_oop) != nullptr) {
+      ThreadsListHandle tlh; // Provides memory barrier
+      if (java_lang_Thread::thread(thread_oop) != nullptr) {
```
Suggestion for comment after this line:
```
// We still have a live JavaThread, now protected by the ThreadsListHandle.

dcubed-ojdk replied:
I've added:
```
      if (java_lang_Thread::thread(thread_oop) != nullptr) {
        // The still live JavaThread is protected by the ThreadsListHandle.
        JavaThread* thr = nullptr;
```

dholmes-ora commented on src/hotspot/share/prims/unsafe.cpp:
```diff
@@ -783,21 +783,23 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
+        // We verified that java_lang_Thread::thread(thread_oop) != nullptr
+        // before and after creating tlh above so quick_mode can be used in
+        // this conversion call.
+        if (tlh.cv_internal_thread_to_JavaThread(jthread, &thr, nullptr, true /* quick_mode */)) {
```
Sorry I'm having a mental blank here now. Why do we even need to call this if we can access `java_lang_Thread::thread(thread_oop)` after the TLH was created?

robehn replied:
As a caller of cv_internal_thread_to_JavaThread we do not know the implementation of quick_mode.

As the implementation is now, we still want the asserts.

dcubed-ojdk replied:
In particular we want this assert from `cv_internal_thread_to_JavaThread()`:
```
  assert(includes(java_thread), "must be");
```
because that verifies our assumptions from this call-site that `quick_mode` is okay.

Also, it is possible that a JavaThread that's racing to exit will make it past `ensure_join()`
after our check on L789 above and run into this block in the conversion function:
```
  JavaThread *java_thread = java_lang_Thread::thread(thread_oop);
  if (java_thread == nullptr) {
    // The java.lang.Thread does not contain a JavaThread* so it has not
    // run enough to be put on a ThreadsList or it has exited enough to
    // make it past ensure_join() where the JavaThread* is cleared.
    return false;
  }
```

Because we (the observer thread) captured the target JavaThread in the
ThreadsListHandle, it's safe for us to use the `JavaThread*`, but I've always
been a fan of bailing on operations like these as soon as we detect that the
`JavaThread*` is exiting.

dholmes-ora replied:
But ignoring the asserts `cv_internal_thread_to_JavaThread()` serves absolutely no purpose when we could request quick-mode - and we don't need to know what `cv_internal_thread_to_JavaThread` does - we know how thread startup and exit is managed and we know what a TLH guarantees . Sure we can bail sooner but there is no need to, so that is not required. In quick-mode, barring the assertion for includes, it all boils down to nothing! We don't need the Thread_oop passed out as we already have it before we start; we don't even need the JavaThread* passed out as we already have it too! So what does it actually do for us? Nothing. We could just have:
```
UNSAFE_ENTRY(void, Unsafe_Unpark(JNIEnv *env, jobject unsafe, jobject jthread)) {
  if (jthread != nullptr) {
    oop thread_oop = JNIHandles::resolve_non_null(jthread);
    if (java_lang_Thread::thread(thread_oop) != nullptr) {
      // Try to capture the live JavaThread in a ThreadsListHandle:
      ThreadsListHandle tlh; // Provides memory barrier
      JavaThread* thr;
      if ( (thr = java_lang_Thread::thread(thread_oop)) != nullptr) {
        assert(tlh.includes(thr), "must be");
        // The still live JavaThread is protected by the ThreadsListHandle
        // so is safe to access.
        Parker* p = thr->parker();
        HOTSPOT_THREAD_UNPARK((uintptr_t) p);
        p->unpark();
      }
    } // ThreadsListHandle is destroyed here.
  }
} UNSAFE_END
```

robehn replied:
It's also possible to create a special version of ThreadsListHandle which takes an Thread object in constructor.
Do the eetop check and then safe fetch the ThreadsList.
If you call include() with the same JavaThread* as the eetop was before creating the ThreadsList, you can return true directly, i.e. quick mode.

```
SpecialThreadsListHandle::SpecialThreadsListHandle(oop thread_oop) {
  _javaThread_from_eetop = java_lang_Thread::thread(thread_oop);
  _threadList = ...
}

bool includes(JavaThread* jt) {
  if (jt == _javaThread_from_eetop && jt != nullptr) return true;
  return _threadsList->includes(jt);
}
```

dcubed-ojdk replied:
Yes, we could create a special version of ThreadsListHandle that would allow us
to encapsulate this special purpose logic in threadsSMR source files where it
belongs. Of course, because I can't stand the 'eetop' name, I'll use some other
naming here. :-)

dcubed-ojdk replied:
Hmmm... it occurs me that this special version of ThreadsListHandle will need a
way to communicate with the call-site that created it and return the protected
`JavaThread*`. If it returns `nullptr`, then the `thread_oop` couldn't be
converted into a protected `JavaThread*`.

Perhaps `stlh.protected_java_thread()`...

dholmes-ora reviewed (commented):
Sorry Dan but it seems to me that if quick_mode is true then we don't even need to call `cv_internal_thread_to_JavaThread` in the first place.

dcubed-ojdk commented:
@apangin - thanks for taking a look at the code. Even as a committer you can chose to
add your review (plus or minus) to a PR. It would still require a (R)eviewer, but I always
count all reviews.

@robehn - Thanks for the review and thanks for chiming in with replies to @dholmes-ora
queries. I may add more to those threads.

@dholmes-ora - Thanks for the review! I'll start going thru your comments shortly.

dcubed-ojdk commented:
v02 has been tested with the usual local tests for this project.
Mach5 Tier1 passed with no failures. I'm doing additional Mach5 testing.

dholmes-ora commented on src/hotspot/share/runtime/globals.hpp:
```diff
@@ -949,9 +949,6 @@ const int ObjectAlignmentInBytes = 8;
           "null (+offset) will not raise a SIGSEGV, i.e.,"                  \
           "ImplicitNullChecks don't work (PPC64).")                         \
                                                                             \
-  product(bool, EnableThreadSMRExtraValidityChecks, true, DIAGNOSTIC,       \
-             "Enable Thread SMR extra validity checks")                     \
```
Do you still want to make this change as part of this PR?

dcubed-ojdk replied:
Yes I would. All the analysis is recorded in this bug and the PR.

dholmes-ora reviewed (commented):
Unpark changes look good. :) Thanks.

dcubed-ojdk commented:
v03 has been tested with the usual local tests for this project.
Mach5 Tier1 passed with no failures. I'm doing additional Mach5 testing.

robehn reviewed (approved):
Thanks! Works for me!

dcubed-ojdk commented:
@robehn - Thanks for the re-review!

dcubed-ojdk commented:
/contributor add @robehn

dcubed-ojdk commented:
@dholmes-ora - Please re-review when you get the cycles.

dholmes-ora commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1565,11 +1565,11 @@ void java_lang_Thread::serialize_offsets(SerializeClosure* f) {
 
 JavaThread* java_lang_Thread::thread(oop java_thread) {
-  return reinterpret_cast<JavaThread*>(java_thread->address_field(_eetop_offset));
+  return reinterpret_cast<JavaThread*>(java_thread->address_field_acquire(_eetop_offset));
 }
 
 void java_lang_Thread::set_thread(oop java_thread, JavaThread* thread) {
-  java_thread->address_field_put(_eetop_offset, (address)thread);
+  java_thread->release_address_field_put(_eetop_offset, (address)thread);
 }
```
There is a convention that when we have acquire/release semantics, that we make that evident in the name of the method so that it is obvious at the call-site. So these would need renaming to `thread_acquire` and `release_set_thread`.

dholmes-ora replied:
Or, to avoid a lot of changes, add new methods for the acquire/release variants. Really only the setting to non-null requires release semantics. Which reads need acquire semantics is much harder to analyse - but most will not due to other inherent synchronization either via Threads_lock, or at the Java level.

dcubed-ojdk replied:
I completely forgot about the `foo_acquire` and `release_set_foo` convention.
I think I'll go the route of adding `thread_acquire` and `release_set_thread` and
then will revisit each call-site to decide which to use.

I'm also going to look at splitting the `java_lang_Thread::thread_acquire` and
`java_lang_Thread::release_set_thread` changes into a sub-task. This will allow
for easier reviewing and backport decisions.

dholmes-ora commented on src/hotspot/share/runtime/javaThread.cpp:
```diff
@@ -738,10 +738,10 @@ static void ensure_join(JavaThread* thread) {
-  OrderAccess::release();
+  // Clear the native thread instance - this makes isAlive return false and
+  // allows the join() to complete once we've done the notify_all() below.
+  // Needs a release() to obey Java Memory Model requirements (which is done
+  // in set_thread()).
```
With the suggested renaming you don't need to add comments like this.

dcubed-ojdk replied:
Most of the comment isn't mine; I just updated it slightly and reformatted it. Is the
mention of "Java Memory Model requirements" not a useful thing to note here?

dholmes-ora replied:
Sorry I only meant your addition to the comment. The rest is important, if somewhat vague. The requirements is that if you call `t.isAlive()` and get false (because it has terminated) then you must see all stores performed by `t` before it terminated. We in part achieve this by marking `eetop` as a volatile field in the j.l.Thread object, hence the VM (technically) should always do a "volatile write" to it (i.e. release semantics). Conversely, reads of `eetop` from within the VM should have volatile-read semantics (i.e. acquire).

dholmes-ora commented on src/hotspot/share/prims/unsafe.cpp:
```diff
@@ -783,30 +783,20 @@ UNSAFE_ENTRY(void, Unsafe_Park(JNIEnv *env, jobject unsafe, jboolean isAbsolute,
+    // Get the JavaThread* stored in the java.lang.Thread object _before_
+    // the embedded ThreadsListHandle is constructed so we know if the
+    // early life stage of the JavaThread* is protected.
+    FastThreadsListHandle ftlh(thread_oop, java_lang_Thread::thread(thread_oop));
+    JavaThread* thr = ftlh.protected_java_thread();
```
I find this obscures things more than it aids understanding in this case. You don't need a special FTLH to get the desired semantics here.

dholmes-ora replied:
Note this is just a comment I'm not requesting you change it back.

dcubed-ojdk replied:
Thanks for making it clear that the above is just a comment.

I rather like the fact that this is best encapsulation of this crazy performance
logic and it allows us to put (most of) the logic over in the threadSMR source
files where it belongs.

Of all the possible solutions we have tried out, I like this one the most and I
think that it's the least hacky. I wouldn't quite say that this solution "sings",
but it comes the closest (from my POV)...

dcubed-ojdk commented:
@dholmes-ora - Thanks for the re-review. I have more changes to make and
I need to checkout the performance testing that Eric Caspole has done with
his new JMH. His JMH reproduces the degradation from JDK8u -> JDK21
with specific linux-aarch64 machines and shows that this patch restores the
performance in JDK21. I need to double check what the testing on other
platforms shows.

dcubed-ojdk commented:
I just rebased this project from jdk-21+18 -> jdk-21+20. I've only built and tested
the rebase on my MBP13 so far.

dcubed-ojdk commented:
The previous version of this patch has been reworked into two dependent patches:
- [JDK-8307067](https://bugs.openjdk.org/browse/JDK-8307067) remove broken EnableThreadSMRExtraValidityChecks option
- https://github.com/openjdk/jdk/pull/13704

and:
- [JDK-8307068](https://bugs.openjdk.org/browse/JDK-8307068) store a JavaThread* in the java.lang.Thread object after the JavaThread* is added to the main ThreadsList
- https://github.com/openjdk/jdk/pull/13723

with this PR now containing just the performance tweak parts of the original fix.

When I refactored the pieces of  [JDK-8307068](https://bugs.openjdk.org/browse/JDK-8307068)
into that dependent PR, I also addressed @dholmes-ora code review comments on that
portion of the previous patch.

This PR has also been updated to use `java_lang_Thread::thread_acquire()` as needed.

I'm doing the usual Mach5 testing on this patch.

dcubed-ojdk commented:
@dholmes-ora - Please re-review this PR now that two pieces have been refactored
out into JDK-8307067 and JDK-830-7068.

@robehn - Any chance you are around for a re-review?
@coleenp - Any chance you can review this part since you already reviewed JDK-8307067?
@apangin - You looked at this fix when it was part of https://github.com/openjdk/jdk/pull/13519. Any chance you can take another look?

robehn reviewed (approved):
I'm good! Thanks!

dholmes-ora reviewed (approved):
I still have reservations about the utility of FastTLH given it requires a load-acquire by the caller but can't check/enforce that. But I can live with this.

One nit below.

Thanks.

dcubed-ojdk commented:
@robehn and @dholmes-ora - Thanks for the re-reviews!

dcubed-ojdk commented on src/hotspot/share/runtime/threadSMR.cpp:
```diff
@@ -839,6 +839,19 @@ bool ThreadsListHandle::cv_internal_thread_to_JavaThread(jobject jthread,
+  if (java_thread != nullptr) {
+    // We captured a non-nullptr JavaThread* before the _tlh was created
+    // so that covers the early life stage of the target JavaThread.
+    _protected_java_thread = java_lang_Thread::thread_acquire(thread_oop);
```
I will have to disagree since the initialization and construction of this helper object
is racing with the target JavaThread's termination code path in `ensure_join()`.

The caller's `thread_acquire()` and the `thread_acquire()` on L847 are racing with
this code block in `ensure_join()`:
```
 // Clear the native thread instance - this makes isAlive return false and allows the join()
  // to complete once we've done the notify_all below. Needs a release() to obey Java Memory Model
  // requirements.
  java_lang_Thread::release_set_thread(threadObj(), nullptr);
  lock.notify_all(thread);
  // Ignore pending exception, since we are exiting anyway
  thread->clear_pending_exception();
```

If the caller's `thread_acquire()` happens before the call to `release_set_thread()`,
then we'll execute the code block controlled by `L844   if (java_thread != nullptr) {`
and we will get to the `thread_acquire()` call on L847. Let's say that the target
JavaThread has now made it past the `release_set_thread()` call so now the
JavaThread* stored in the java.lang.Thread object is now nullptr. We want the
`thread_acquire()` on L847 to match up with the `release_set_thread()` so that
we see the fact that the JavaThread* stored in the java.lang.Thread object is now
nullptr.

Please let me know if I missed something here.

dholmes-ora replied:
That release in `ensure_join` is for Java volatile semantics for threads calling `is_alive()`. It is there to ensure that any Java writes made by the thread before it terminated are seen by the thread that sees `is_alive()` return false. The VM code we are dealing with here does not need to see any such writes by the terminating thread, nor are there any interesting writes in the VM code which must be seen if `thread()` returned null. The code we are dealing with here only needs to synchronize with the thread startup logic and in particular its addition to the ThreadsList. The `thread_acquire` in the caller ensures that if we see a non-null value then we also see the writes to the ThreadsList as required. The subsequent re-reading of that value will return null or non-null but it does not need to synchronize with any other memory writes, hence acquire semantics are not needed.

dcubed-ojdk replied:
I'm sorry, but I will have to disagree again. Here's the first `thread_acquire()` call
along with the comment that explains that this first call is for the early life stage of
the target JavaThread:
```
    // Get the JavaThread* stored in the java.lang.Thread object _before_
    // the embedded ThreadsListHandle is constructed so we know if the
    // early life stage of the JavaThread* is protected.
    FastThreadsListHandle ftlh(thread_oop, java_lang_Thread::thread_acquire(thread_oop));
```

At this point in the code, all we have done is potentially gather the `JavaThread*`
that's stored in the java.lang.Thread object. That value we gathered, might be
nullptr and it might be non-nullptr. For the first `thread_acquire` call, I see it
as matching up with the `release_set_thread` calls at the end of thread startup.

Here's the helper's constructor code:
```
FastThreadsListHandle::FastThreadsListHandle(oop thread_oop, JavaThread* java_thread) : _protected_java_thread(nullptr) {
  assert(thread_oop != nullptr, "must be");
  if (java_thread != nullptr) {
    // We captured a non-nullptr JavaThread* before the _tlh was created
    // so that covers the early life stage of the target JavaThread.
    _protected_java_thread = java_lang_Thread::thread_acquire(thread_oop);
```

We check the `java_thread` parameter and if it's non-nullptr, then we know
that the `JavaThread*` stored in the java.lang.Thread object was non-nullptr
before we constructed the embedded ThreadsListHandle (_tlh). Now we can
check the second half of the question...

Here's the second `thread_acquire()` call again along with the assert() and the
comment that explains that this second call is for end life stage of the target
`JavaThread*`.
```
    _protected_java_thread = java_lang_Thread::thread_acquire(thread_oop);
    assert(_protected_java_thread == nullptr || _tlh.includes(_protected_java_thread), "must be");
    // If we captured a non-nullptr JavaThread* after the _tlh was created
    // then that covers the end life stage of the target JavaThread and we
    // we know that _tlh protects the JavaThread*.
```

> The subsequent re-reading of that value will return null or non-null but it does
> not need to synchronize with any other memory writes

The second `thread_acquire` most certainly does need to synchronize with the
code in `ensure_join()` that does a `release_set_thread` of a nullptr. If that code
set a nullptr before we could read a non-nullptr value a second time, then we can
make no inferences about what `_tlh` contains. So I see the second `thread_acquire`
as matching up with the `release_set_thread` in `ensure_join`.

If the `_protected_java_thread` field is nullptr, then we don't know that `_tlh` has
captured the target `JavaThread*` and the `JavaThread*` is NOT protected.
If `_protected_java_thread` != nullptr, then we do know that `_tlh` contains the
target `JavaThread*` and `JavaThread*` is protected.

dholmes-ora replied:
> The second `thread_acquire` most certainly does need to synchronize with the
> code in `ensure_join()` that does a `release_set_thread` of a nullptr. If that code
> set a nullptr before we could read a non-nullptr value a second time, then we can
> make no inferences about what `_tlh` contains. So I see the second thread_acquire
> as matching up with the release_set_thread in ensure_join.

The only memory value of interest is the read of eetop. The release/acquire have no affect on the reading or writing of that field itself. We would only need acquire semantics when reading the potentially null eetop if we were then to read another memory value that would be written prior to eetop being set to null. That is not the case hence the acquire semantics are unnecessary.

dcubed-ojdk replied:
I've read thru this thread of comments three times and I think I see where and how my
brain got confused by what we are doing here. Let me see if I can unwind this mess that
I have made...

Our new `FastThreadsListHandle` helper object is used by `Unsafe_Unpark()` to quickly
determine whether our target thread (`jthread` parameter) is protected by the embedded
`_tlh` in the helper object. The `Unsafe_Unpark()` call is racing against two different
situations in the target thread's life-cycle:

1) The first situation is the early life stage of the target thread. It is during this stage that
    the target `JavaThread*` gets added to the main ThreadsList AND the `JavaThread*`
    gets stored in the java.lang.Thread object.

    From the `Unsafe_Unpark()` callers POV, we need to see two things happen: the target
    `JavaThread*` getting added to the main ThreadsList and a non-nullptr value being
    stored in the java.lang.Thread object. Seeing both memory operations requires that we
    use `java_lang_Thread::thread_acquire()` for the query in `Unsafe_Unpark()`.

2) The second situation is the end life stage of the target thread. It is during this stage that
     `nullptr` gets stored in the java.lang.Thread object. The target `JavaThread*` will also
     get removed from the main ThreadsList later, but we don't care about that part since the
     setting of the `nullptr` in the java.lang.Thread object is a perfect bailout sentinel.

     From the `FastThreadsListHandle` helper object's POV, we only care about whether we
     see a non-`nullptr` value or a `nullptr` value. Since we only care about a single memory
     operation, we use `java_lang_Thread::thread()` which is implemented with an underlying
     atomic load operation. Acquire semantics are not needed for this query.

Okay so those are the two situations of interest in the target thread. Now let's talk about
`Unsafe_Unpark()`, its use of the `FastThreadsListHandle` helper object and how that
applies to the target thread.

1) The first query of the `JavaThread*` in the java.lang.Thread object that is made in
    `Unsafe_Unpark()` is trying to determine if the target thread is past the early life stage.
    A non-`nullptr` value means that the `_tlh` embedded in the FastThreadsListHandle
    _might_ be protecting the `JavaThread*`. Since this code is primarily targeted at the
    early life stage, we have to use `java_lang_Thread::thread_acquire()` in order to get
    a proper answer from that code path in the target thread.

    Of course, this code might also execute when the target thread is in the end life stage,
    but we'll still get a valid answer from it so no harm, no foul.

2) `Unsafe_Unpark()` passes that first `JavaThread*` value to the FastThreadsListHandle
     helper object and that constructor does the remainder of the work if the value passed
     in is non-`nullptr`. The constructor does the second query of the `JavaThread*` in the
     java.lang.Thread object and it only cares about whether we see a `nullptr` value or a
     non-`nullptr` value so we can use just `java_lang_Thread::thread()` here to fetch the
     value. A non-`nullptr` value means that the embedded `_tlh` does protect the
     `JavaThread*` and `Unsafe_Unpark()` can proceed to use that `JavaThread*` to do
     its work. A `nullptr` value means that the target `JavaThread*` has exited enough to
     make it past ensure_join() where the `JavaThread*` is cleared from the
     java.lang.Thread object.

     This second query does not execute if the target thread is in the early life stage and a
     `nullptr` value is returned by the first query so there is no opportunity for the
     `java_lang_Thread::thread()` query to return an invalid answer from an early life stage.

dcubed-ojdk commented:
dholmes CR - FastThreadsListHandle ctr should use java_lang_Thread::thread() instead of java_lang_Thread::thread_acquire(); clarify comments.

Passes local build and the usual local tests on my MBP13.

dholmes-ora reviewed (approved):
Thanks Dan - looks good.

dcubed-ojdk commented:
@dholmes-ora - Thanks for the re-review and approval.

dcubed-ojdk commented:
/integrate
